pkg/services/alerting/conditions/query.go

```diff
@@ -108,6 +110,51 @@ func (c *QueryCondition) Eval(context *alerting.EvalContext, requestHandler plug
 	}, nil
 }
 
+func getMinInterval(model *simplejson.Json, dsInfo *models.DataSource) (time.Duration, error) {
+	minIntervalString := ""
+
+	// if panel-min-interval exists, that is the min-interval we use.
+	// otherwise we use datasource-min-interval (if it exists)
+	panelMinIntervalString := model.Get("interval").MustString("")
+	if panelMinIntervalString != "" {
+		minIntervalString = panelMinIntervalString
+	} else {
+		dataSourceMinIntervalString := dsInfo.JsonData.Get("timeInterval").MustString("")
+		if dataSourceMinIntervalString != "" {
+			minIntervalString = dataSourceMinIntervalString
+		}
+	}
+
+	if minIntervalString == "" {
+		return time.Duration(0), nil
+	}
+
+	return tsdb.MinIntervalStringToDuration(minIntervalString)
+}
+
+func calculateInterval(timeRange plugins.DataTimeRange, model *simplejson.Json, dsInfo *models.DataSource) (time.Duration, error) {
+
+	minInterval, err := getMinInterval(model, dsInfo)
+
+	if err != nil {
+		return time.Duration(0), err
+	}
+
+	calc := tsdb.NewCalculator()
+
+	newFormatTimeRange := backend.TimeRange{
+		From: timeRange.MustGetFrom(),
+		To:   timeRange.MustGetTo(),
+	}
+
+	interval, err := calc.Calculate(newFormatTimeRange, minInterval, tsdb.Min)
```

Comment:
we call into `tsdb.Calculate` to make the code simpler, but we could have here a simplified copy of `tsdb.Calculate` in like 4lines of code, i wonder if we should do that.